firedrake/slate/slac/compiler.py

```diff
@@ -70,154 +223,255 @@ def compile_expression(slate_expr, tsfc_parameters=None):
-
-    # Initialize coordinate, cell orientations and facet/layer
-    # symbols
-    coordsym = ast.Symbol("coords")
-    coords = None
-    cell_orientations = ast.Symbol("cell_orientations")
-    cellfacetsym = ast.Symbol("cell_facets")
-    mesh_layer_sym = ast.Symbol("layer")
-    inc = []
-
-    # We keep track of temporaries that have been declared
+    builder = LocalKernelBuilder(expression=slate_expr,
+                                 tsfc_parameters=tsfc_parameters)
+
     declared_temps = {}
-    for cxt_kernel in builder.context_kernels:
-        exp = cxt_kernel.tensor
+    statements = [ast.FlatBlock("/* Declare and initialize */\n")]
+    for exp in builder.temps:
+        # Declare and initialize terminal temporaries
         t = builder.temps[exp]
+        statements.append(ast.Decl(eigen_matrixbase_type(exp.shape), t))
+        statements.append(ast.FlatBlock("%s.setZero();\n" % t))
+        declared_temps[exp] = t
 
-        if exp not in declared_temps:
-            # Declare and initialize the temporary
-            statements.append(ast.Decl(eigen_matrixbase_type(exp.shape), t))
-            statements.append(ast.FlatBlock("%s.setZero();\n" % t))
-            declared_temps[exp] = t
+    statements.append(ast.FlatBlock("/* Assemble local tensors */\n"))
 
-        it_type = cxt_kernel.original_integral_type
+    # Cell integrals are straightforward. Just splat them out.
+    statements.extend(builder.assembly_calls["cell"])
 
-        if it_type not in supported_integral_types:
-            raise NotImplementedError("Type %s not supported." % it_type)
-
-        # Explicit checking of coordinates
-        coordinates = exp.ufl_domain().coordinates
-        if coords is not None:
-            assert coordinates == coords
+    if builder.needs_cell_facets:
+        # The for-loop will have the general structure:
+        #
+        #    FOR (facet=0; facet<num_facets; facet++):
+        #        IF (facet is interior):
+        #            *interior calls
+        #        ELSE IF (facet is exterior):
+        #            *exterior calls
+        #
+        # If only interior (exterior) facets are present,
+        # then only a single IF-statement checking for interior
+        # (exterior) facets will be present within the loop. The
+        # cell facets are labelled `1` for interior, and `0` for
+        # exterior.
+
+        statements.append(ast.FlatBlock("/* Loop over cell facets */\n"))
+        chker = {"interior_facet": 1,
+                 "exterior_facet": 0,
+                 "interior_facet_vert": 1,
+                 "exterior_facet_vert": 0}
+        int_calls = list(chain(*[builder.assembly_calls[it_type]
+                                 for it_type in chker
+                                 if chker[it_type] == 1]))
+        ext_calls = list(chain(*[builder.assembly_calls[it_type]
+                                 for it_type in chker
+                                 if chker[it_type] == 0]))
```

Comment:
Ugh, prefer `if it_type in set(...)`.

Reply:
You mean something like:
```python
int_calls = list(chain(*[builder.assembly_calls[it_type]
                         for it_type in ("interior_facet", "interior_facet_vert")]))
```
?

Reply:
Oh, I guess yes, I hadn't realised you were iterating over the integral types in chker and filtering based on the value.